File: swift/Sources/Schemas/OpenInferenceSpan.swift
import Foundation

public struct OpenInferenceSpan: Codable, Hashable, Sendable {
    public let attributes: [String: JSONValue]?
    public let endTime: Nullable<String>?
    public let name: Nullable<String>?
    public let openinferenceKind: Nullable<String>?
    public let parentSpanId: Nullable<String>?
    public let raw: [String: JSONValue]?
    public let resourceAttributes: [String: JSONValue]?
    public let spanId: String
    public let startTime: Nullable<String>?
    public let statusCode: Nullable<String>?
    public let statusMessage: Nullable<String>?
    public let traceId: String
    /// Additional properties that are not explicitly defined in the schema
    public let additionalProperties: [String: JSONValue]

    public init(
        attributes: [String: JSONValue]? = nil,
        endTime: Nullable<String>? = nil,
        name: Nullable<String>? = nil,
        openinferenceKind: Nullable<String>? = nil,
        parentSpanId: Nullable<String>? = nil,
        raw: [String: JSONValue]? = nil,
        resourceAttributes: [String: JSONValue]? = nil,
        spanId: String,
        startTime: Nullable<String>? = nil,
        statusCode: Nullable<String>? = nil,
        statusMessage: Nullable<String>? = nil,
        traceId: String,
        additionalProperties: [String: JSONValue] = .init()
    ) {
        self.attributes = attributes
        self.endTime = endTime
        self.name = name
        self.openinferenceKind = openinferenceKind
        self.parentSpanId = parentSpanId
        self.raw = raw
        self.resourceAttributes = resourceAttributes
        self.spanId = spanId
        self.startTime = startTime
        self.statusCode = statusCode
        self.statusMessage = statusMessage
        self.traceId = traceId
        self.additionalProperties = additionalProperties
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self.attributes = try container.decodeIfPresent([String: JSONValue].self, forKey: .attributes)
        self.endTime = try container.decodeNullableIfPresent(String.self, forKey: .endTime)
        self.name = try container.decodeNullableIfPresent(String.self, forKey: .name)
        self.openinferenceKind = try container.decodeNullableIfPresent(String.self, forKey: .openinferenceKind)
        self.parentSpanId = try container.decodeNullableIfPresent(String.self, forKey: .parentSpanId)
        self.raw = try container.decodeIfPresent([String: JSONValue].self, forKey: .raw)
        self.resourceAttributes = try container.decodeIfPresent([String: JSONValue].self, forKey: .resourceAttributes)
        self.spanId = try container.decode(String.self, forKey: .spanId)
        self.startTime = try container.decodeNullableIfPresent(String.self, forKey: .startTime)
        self.statusCode = try container.decodeNullableIfPresent(String.self, forKey: .statusCode)
        self.statusMessage = try container.decodeNullableIfPresent(String.self, forKey: .statusMessage)
        self.traceId = try container.decode(String.self, forKey: .traceId)
        self.additionalProperties = try decoder.decodeAdditionalProperties(using: CodingKeys.self)
    }

    public func encode(to encoder: Encoder) throws -> Void {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try encoder.encodeAdditionalProperties(self.additionalProperties)
        try container.encodeIfPresent(self.attributes, forKey: .attributes)
        try container.encodeNullableIfPresent(self.endTime, forKey: .endTime)
        try container.encodeNullableIfPresent(self.name, forKey: .name)
        try container.encodeNullableIfPresent(self.openinferenceKind, forKey: .openinferenceKind)
        try container.encodeNullableIfPresent(self.parentSpanId, forKey: .parentSpanId)
        try container.encodeIfPresent(self.raw, forKey: .raw)
        try container.encodeIfPresent(self.resourceAttributes, forKey: .resourceAttributes)
        try container.encode(self.spanId, forKey: .spanId)
        try container.encodeNullableIfPresent(self.startTime, forKey: .startTime)
        try container.encodeNullableIfPresent(self.statusCode, forKey: .statusCode)
        try container.encodeNullableIfPresent(self.statusMessage, forKey: .statusMessage)
        try container.encode(self.traceId, forKey: .traceId)
    }

    /// Keys for encoding/decoding struct properties.
    enum CodingKeys: String, CodingKey, CaseIterable {
        case attributes
        case endTime = "end_time"
        case name
        case openinferenceKind = "openinference_kind"
        case parentSpanId = "parent_span_id"
        case raw
        case resourceAttributes = "resource_attributes"
        case spanId = "span_id"
        case startTime = "start_time"
        case statusCode = "status_code"
        case statusMessage = "status_message"
        case traceId = "trace_id"
    }
}

File: rust/src/api/types/open_inference_span.rs
pub use crate::prelude::*;

#[derive(Debug, Clone, Serialize, Deserialize, Default, PartialEq)]
pub struct OpenInferenceSpan {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub attributes: Option<HashMap<String, serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub end_time: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub openinference_kind: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parent_span_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub raw: Option<HashMap<String, serde_json::Value>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resource_attributes: Option<HashMap<String, serde_json::Value>>,
    #[serde(default)]
    pub span_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub start_time: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status_message: Option<String>,
    #[serde(default)]
    pub trace_id: String,
    /// Additional properties that are not part of the defined schema.
    #[serde(flatten)]
    pub extra: std::collections::HashMap<String, serde_json::Value>,
}

impl OpenInferenceSpan {
    pub fn builder() -> OpenInferenceSpanBuilder {
        <OpenInferenceSpanBuilder as Default>::default()
    }
}

#[derive(Clone, PartialEq, Default, Debug)]
#[non_exhaustive]
pub struct OpenInferenceSpanBuilder {
    attributes: Option<HashMap<String, serde_json::Value>>,
    end_time: Option<String>,
    name: Option<String>,
    openinference_kind: Option<String>,
    parent_span_id: Option<String>,
    raw: Option<HashMap<String, serde_json::Value>>,
    resource_attributes: Option<HashMap<String, serde_json::Value>>,
    span_id: Option<String>,
    start_time: Option<String>,
    status_code: Option<String>,
    status_message: Option<String>,
    trace_id: Option<String>,
}

impl OpenInferenceSpanBuilder {
    pub fn attributes(mut self, value: HashMap<String, serde_json::Value>) -> Self {
        self.attributes = Some(value);
        self
    }

    pub fn end_time(mut self, value: impl Into<String>) -> Self {
        self.end_time = Some(value.into());
        self
    }

    pub fn name(mut self, value: impl Into<String>) -> Self {
        self.name = Some(value.into());
        self
    }

    pub fn openinference_kind(mut self, value: impl Into<String>) -> Self {
        self.openinference_kind = Some(value.into());
        self
    }

    pub fn parent_span_id(mut self, value: impl Into<String>) -> Self {
        self.parent_span_id = Some(value.into());
        self
    }

    pub fn raw(mut self, value: HashMap<String, serde_json::Value>) -> Self {
        self.raw = Some(value);
        self
    }

    pub fn resource_attributes(mut self, value: HashMap<String, serde_json::Value>) -> Self {
        self.resource_attributes = Some(value);
        self
    }

    pub fn span_id(mut self, value: impl Into<String>) -> Self {
        self.span_id = Some(value.into());
        self
    }

    pub fn start_time(mut self, value: impl Into<String>) -> Self {
        self.start_time = Some(value.into());
        self
    }

    pub fn status_code(mut self, value: impl Into<String>) -> Self {
        self.status_code = Some(value.into());
        self
    }

    pub fn status_message(mut self, value: impl Into<String>) -> Self {
        self.status_message = Some(value.into());
        self
    }

    pub fn trace_id(mut self, value: impl Into<String>) -> Self {
        self.trace_id = Some(value.into());
        self
    }

    /// Consumes the builder and constructs a [`OpenInferenceSpan`].
    /// This method will fail if any of the following fields are not set:
    /// - [`span_id`](OpenInferenceSpanBuilder::span_id)
    /// - [`trace_id`](OpenInferenceSpanBuilder::trace_id)
    pub fn build(self) -> Result<OpenInferenceSpan, BuildError> {
        Ok(OpenInferenceSpan {
            attributes: self.attributes,
            end_time: self.end_time,
            name: self.name,
            openinference_kind: self.openinference_kind,
            parent_span_id: self.parent_span_id,
            raw: self.raw,
            resource_attributes: self.resource_attributes,
            span_id: self
                .span_id
                .ok_or_else(|| BuildError::missing_field("span_id"))?,
            start_time: self.start_time,
            status_code: self.status_code,
            status_message: self.status_message,
            trace_id: self
                .trace_id
                .ok_or_else(|| BuildError::missing_field("trace_id"))?,
            extra: Default::default(),
        })
    }
}
